mlflow/tensorflow.py

```diff
@@ -619,12 +620,41 @@ def _flush_queue():
         # flush operation should proceed; all others are redundant and should be dropped
         acquired_lock = _metric_queue_lock.acquire(blocking=False)
         if acquired_lock:
-            global _metric_queue
             client = mlflow.tracking.MlflowClient()
-            dic = _assoc_list_to_map(_metric_queue)
-            for key in dic:
-                try_mlflow_log(client.log_batch, key, metrics=dic[key], params=[], tags=[])
-            _metric_queue = []
+            # For thread safety and to avoid modifying a list while iterating over it, we record a
+            # separate list of the items being flushed and remove each one from the metric queue,
+            # rather than clearing the metric queue or reassigning it (clearing / reassigning is
+            # dangerous because we don't block threads from adding to the queue while a flush is
+            # in progress)
+            snapshot = _metric_queue[:]
+            for item in snapshot:
+                _metric_queue.remove(item)
+
+            # NB: Most TensorFlow APIs use one-indexing for epochs, while tf.Keras
+            # uses zero-indexing. Accordingly, the modular arithmetic used here is slightly
+            # different from the arithmetic used in `__MLflowTfKeras2Callback.on_epoch_end`,
+            # which provides metric logging hooks for tf.Keras
+            metrics_to_log_immediately = [
+                item for item in snapshot if (item[1].step - 1) % _LOG_EVERY_N_STEPS == 0
+            ]
+
+            metrics_by_run = _assoc_list_to_map(metrics_to_log_immediately)
+            for run_id, metrics in metrics_by_run.items():
+                try_mlflow_log(client.log_batch, run_id, metrics=metrics, params=[], tags=[])
+
+            # If metrics for the last training epoch were not logged due to filtering on
+            # `every_n_iter`, log them now
```

Comment:
I'm a little confused, why exactly is this necessary? If they were not logged due to filtering on `every_n_iter`, isn't that the purpose of that filter?

Reply:
That's a fair point. I was concerned that users might miss metrics for the final training epoch, which is more likely to produce the best results compared to other randomly selected epochs. However, because we're defaulting to `every_n_iter=1`, it seems acceptable to drop metrics for the last epoch for cases where `every_n_iter != 1` since users are opting into this behavior & likely know what they're doing. I've removed the logic that was introduced to unconditionally log metrics for the last epoch, which simplifies things quite a bit. Thanks!

Reply:
Sounds great!